featurebyte/service/use_case.py

```diff
@@ -0,0 +1,190 @@
+        persistent: Persistent,
+        catalog_id: Optional[ObjectId],
+        observation_table_service: ObservationTableService,
+        context_service: ContextService,
+        historical_feature_table_service: HistoricalFeatureTableService,
+    ):
+        super().__init__(user, persistent, catalog_id)
+        self.observation_table_service = observation_table_service
+        self.context_service = context_service
+        self.historical_feature_table_service = historical_feature_table_service
+
+    async def create_use_case(self, data: UseCaseCreate) -> UseCaseModel:
+        """
+        Create a UseCaseModel document
+
+        Parameters
+        ----------
+        data: UseCaseCreate
+            use case creation data
+
+        Returns
+        -------
+        UseCaseModel
+        """
+        data = UseCaseCreateTarget(**data.dict(), _id=data.id)
+
+        # automatically associate observation tables with the use case
+        # if they have the same context_id and target_id
+        obs_table_ids = []
+        async for obs_table in self.observation_table_service.list_documents_iterator(
+            query_filter={"context_id": data.context_id, "request_input.target_id": data.target_id},
+        ):
+            obs_table_ids.append(obs_table.id)
+        data.observation_table_ids = obs_table_ids
+
+        return await self.create_document(data=data)
+
+    async def update_use_case(
+        self,
+        document_id: ObjectId,
+        data: UseCaseUpdate,
+    ) -> UseCaseModel:
+        """
+        Update a UseCaseModel document
+
+        Parameters
+        ----------
+        document_id: ObjectId
+            use case id
+        data: UseCaseUpdate
+            use case update data
+
+        Returns
+        -------
+        UseCaseModel
+        """
+        use_case: UseCaseModel = await self.get_document(document_id=document_id)
+        data = UseCaseUpdateTarget(
+            **data.dict(), observation_table_ids=use_case.observation_table_ids
+        )
+
+        if data.default_preview_table_id:
+            # validate and add default_preview_table_id
+            await self._validate_input_observation_table(use_case, data.default_preview_table_id)
+            data.observation_table_ids.append(data.default_preview_table_id)
+
+        if data.default_eda_table_id:
+            # validate and add default_eda_table_id
+            await self._validate_input_observation_table(use_case, data.default_eda_table_id)
+            data.observation_table_ids.append(data.default_eda_table_id)
+
+        if data.new_observation_table_id:
+            # validate and add new_observation_table_id
+            await self._validate_input_observation_table(use_case, data.new_observation_table_id)
+            data.observation_table_ids.append(data.new_observation_table_id)
+
+        # remove duplicate observation_table_ids
+        data.observation_table_ids = list(set(data.observation_table_ids))
```

Comment:
Note that observation table can be deleted. Just to remind that we may need to block user to delete observation table if it is referenced in any use case.